L1Trigger/TrackFindingTracklet/src/KFin.cc

```diff
@@ -0,0 +1,418 @@
+                 baseUphiT_);
+        const double phi0S = digi(phi0U - setup_->hybridRangePhi() / 2., baseUphiT_);
+        const double cot = digi(ttTrackRef->tanL(), baseUcot_);
+        const double z0 = digi(ttTrackRef->z0(), baseUzT_);
+        const double phiT = digi(phi0S + r2Inv * digi(dataFormats_->chosenRofPhi(), baseUr_), baseUphiT_);
+        const double zT = digi(z0 + cot * digi(setup_->chosenRofZ(), baseUr_), baseUzT_);
+        // kill tracks outside of fiducial range
+        if (abs(phiT) > setup_->baseRegion() / 2. || abs(zT) > setup_->hybridMaxCot() * setup_->chosenRofZ() ||
+            abs(z0) > setup_->beamWindowZ()) {
+          input.push_back(nullptr);
+          continue;
+        }
+        // convert stubs
+        vector<Stub*> stubs;
+        stubs.reserve(channelAssignment_->numProjectionLayers(channel) + channelAssignment_->numSeedingLayers());
+        for (int layer = 0; layer < channelAssignment_->numProjectionLayers(channel); layer++) {
+          const FrameStub& frameStub = streamsStub[offsetStub + layer][frame];
+          const TTStubRef& ttStubRef = frameStub.first;
+          const int layerId = channelAssignment_->layerId(channel, layer);
+          if (ttStubRef.isNull())
+            continue;
+          // parse residuals from tt::Frame and take r and layerId from tt::TTStubRef
+          const bool barrel = setup_->barrel(ttStubRef);
+          const int layerIdTracklet = channelAssignment_->trackletLayerId(ttStubRef);
+          const double basePhi = barrel ? settings_->kphi1() : settings_->kphi(layerIdTracklet);
+          const double baseRZ = barrel ? settings_->kz(layerIdTracklet) : settings_->kz();
+          const int widthRZ = barrel ? setup_->tbWidthZ() : setup_->tbWidthR();
+          TTBV hw(frameStub.second);
+          const TTBV hwRZ(hw, widthRZ, 0, true);
+          hw >>= widthRZ;
+          const TTBV hwPhi(hw, setup_->tbWidthPhi(), 0, true);
+          hw >>= setup_->tbWidthPhi();
+          const double r = digi(setup_->stubR(hw, ttStubRef) - dataFormats_->chosenRofPhi(), baseUr_);
+          double phi = hwPhi.val(basePhi);
+          if (basePhi > baseUphi_)
+            phi += baseUphi_ / 2.;
+          const double z = digi(hwRZ.val(baseRZ) * (barrel ? 1. : -cot), baseUz_);
+          // determine module type
+          double rz = r + digi(dataFormats_->chosenRofPhi(), baseUr_);
+          if (barrel)
+            rz = rz * cot + z0 + z;
+          const int indexLayerId = setup_->indexLayerId(ttStubRef);
+          const double limit = barrel ? setup_->tiltedLayerLimitZ(indexLayerId) : setup_->psDiskLimitR(indexLayerId);
+          const bool psTilt = barrel ? (abs(rz) < limit) : setup_->psModule(ttStubRef);
+          stubs_.emplace_back(ttStubRef, layerId, r, phi, z, psTilt);
+          stubs.push_back(&stubs_.back());
+        }
+        // create fake seed stubs
```

Comment:
Expand comment "Create fake seed stubs, since TrackBuilder doesn't output these stubs", required by the KF.

Reply:
done